src/backups.py

```diff
@@ -0,0 +1,301 @@
+        logger.info("Checking if cluster is in blocked state")
+        if self.charm._is_cluster_blocked():
+            return False, "Cluster or unit is in a blocking state"
+
+        logger.info("Checking state and role of unit")
+
+        try:
+            state, role = self.charm._mysql.get_member_state()
+        except MySQLGetMemberStateError:
+            return False, "Error obtaining member state"
+
+        if role == "primary" and self.charm.app.planned_units() > 1:
+            return False, "Unit cannot perform backups as it is the cluster primary"
+
+        if state in ["recovering", "offline", "error"]:
+            return False, f"Unit cannot perform backups as its state is {state}"
+
+        return True, None
+
+    def _pre_backup(self) -> Tuple[bool, str]:
+        """Runs operations required before performing a backup."""
+        logger.info("Setting cluster state as 'backing-up'")
+        self.charm.unit_peer_data["cluster-state"] = "backing-up"
+
+        try:
+            logger.info("Setting unit option tag:_hidden")
+            self.charm._mysql.set_instance_option("tag:_hidden", "true")
+        except MySQLSetInstanceOptionError:
+            self.charm.unit_peer_data["cluster-state"] = "active"
+            return False, "Error setting instance option tag:_hidden"
+
+        try:
+            logger.info("Setting unit as offline before performing backup")
+            self.charm._mysql.set_instance_offline_mode(True)
+        except MySQLSetInstanceOfflineModeError:
+            self.charm.unit_peer_data["cluster-state"] = "active"
+            self.charm._mysql.set_instance_option("tag:_hidden", "false")
+            return False, "Error setting instance as offline before performing backup"
+
+        return True, None
+
+    def _upload_logs_to_s3(
+        self: str,
+        stdout: str,
+        stderr: str,
+        s3_bucket: str,
+        log_filename: str,
+        s3_region: str,
+        s3_endpoint: str,
+        s3_access_key: str,
+        s3_secret_key: str,
+    ) -> bool:
+        logs = f"""Stdout:
+{stdout}
+
+Stderr:
+{stderr}
+        """
+        logger.debug(f"Output of xtrabackup: {logs}")
+
+        logger.info("Uploading output of xtrabackup to S3")
+        return upload_content_to_s3(
+            logs,
+            s3_bucket,
+            log_filename,
+            s3_region,
+            s3_endpoint,
+            s3_access_key,
+            s3_secret_key,
+        )
+
+    def _backup(
+        self,
+        s3_bucket: str,
+        s3_directory: str,
+        s3_endpoint: str,
+        s3_region: str,
+        s3_access_key: str,
+        s3_secret_key: str,
+    ) -> None:
+        """Runs the backup operations."""
+        try:
+            logger.info("Running the xtrabackup commands")
+            stdout, stderr = self.charm._mysql.execute_backup_commands(
+                s3_bucket,
+                f"{s3_directory}/backup",
```

Comment:
why do we need to create an additional folder?
```suggestion
                f"{s3_directory}",
```

Reply:
addressed in aedeff6b9921c57431722fcf5a827a4509226ae3